### ISSUE TITLE
Automatically generate traceability graphs

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#801]
As mentioned in the comments of commit 2600f0a, the note that was present in the introduction for traceability graphs

> NOTE: Building a tool to automatically generate the graphical representation of the matrix by scanning the labels and reference can be future work.

is related to Drasil more than it's related to an SRS' content.

This issue is being made to ensure that the removal of the sentence from the current implementation doesn't lead to the future-work task being ignored.

**Comment by smiths** [comment]
I see above that #1537 is "scheduled" to close this issue.  I disagree.  #1537 creates traceability matrices.  Creating traceability graphs is a separate task to help us visualize the traceability.  @Mornix, assuming you agree, can you please "unlink" this issue from #1537.

@Mornix and @JacquesCarette, would generating the traceability graph be feasible enough a task that one of the summer students could be assigned to it?  At one time @palmerst was generating traceability graphs (or maybe it was uses graphs?), so it seems like there are tools in place to do it.  With the updates from @Mornix the graphs can be more manageable.

**Comment by JacquesCarette** [comment]
The traceability matrices are incidence matrices already, so getting a graph out of that should be quite easy. Or in the same way (i.e. following the dependency relations).  In either way, then generating `.dot` files that can be turned into various other formats in the build is most doable.

**Comment by Mornix** [comment]
Reopening this because I missed untagging it before #1537 was merged.

**Comment by JacquesCarette** [comment]
And this is fixed, as that was merged.

**Comment by smiths** [comment]
@JacquesCarette, we have traceability matrices, but not traceability graphs.  @Mornix reopened this issue because this issue was not actually supposed to be tied to #1537.  It isn't the highest priority, but I think a traceability graph might provide us with some insights that are hard to visualize from the matrix.

I'm going to reopen this issue, because, as far as I know, we don't have traceability graphs.

**Comment by Nathaniel-Hu** [comment]
@smiths and @JacquesCarette, 
I have created a sample .dot file (and .png file) for the first traceability matrix found in the SWHS Drasil example SRS. I have uploaded both files to my folder in the People folder (/Drasil/People/Nathaniel/) in the Nathaniel-DataTable branch. Could you review the .dot/.png files to ensure that the formatting of the sample traceability graph is correct? Thanks!

**Comment by JacquesCarette** [comment]
I will let @smiths review the details.

What you should be thinking about is how you would generate that `.dot` file from Drasil. What information would a program need to write that graph?

**Comment by Nathaniel-Hu** [comment]
Hi @JacquesCarette, Is it safe for me to refer to the following file (is it up to date)? I am currently using this file as a basis to extract the assumptions data, as well as extract the dependencies for each assumption.

https://github.com/JacquesCarette/Drasil/blob/2600f0a40611325b41791228843db184a5c2e012/code/drasil-example/Drasil/SWHS/Assumptions.hs

**Comment by JacquesCarette** [comment]
Yes. If the file is in the code base (and used), then it is up to date. [It could still have mistakes, but that's a different matter.]

**Comment by Nathaniel-Hu** [comment]
Ok then. Thank you so much!

**Comment by smiths** [comment]
@Nathaniel-Hu, I have looked at your generated graph (in your folder).  Great start!  If possible, I would like the graph to be changed in the following ways:

- The assumptions should be on the bottom, not on the top
- The separate graphs (the graphs that are disconnected from each other) should be displayed vertically, rather than horizontally.  This would make the graph less wide, and would facilitate making it easier to read

There are also pieces of the graph missing.  For instance, TM: nwtnCooling is referenced by GD: htFluxWaterFromCoil and GD: htFluxPCMFromWater.  This relationship is missing in your diagram.

**Comment by Nathaniel-Hu** [comment]
@smiths, for the generated traceability graphs, should I include all of the relationships across all components in one giant graph? Or is there a specific way that you would like me to split up the relationships across several graphs?

Also, for the Haskell script to be used to generate these traceability graphs: should I model it as code that can be inserted into an existing file? Or should I model it as a standalone script that will be imported into another script?

**Comment by JacquesCarette** [comment]
My opinion is that giant graphs are unreadable. If there is any meaningful way to split it into smaller graphs, that's worth investigating.

The haskell code should be done so that it can be integrated straight into Drasil. It can be a new file, if you wish.

**Comment by smiths** [comment]
I agree with @JacquesCarette that giant graphs are unreadable.  However, as an intermediate step, it might make sense to generate the entire graph, and then use this to visualize how to split it up.  The initial graph was partially disconnected.  When this occurs, we have a natural means to split up the graph.

I think the graph is going to deepen our understanding of the examples.  Visualizing what is going on is difficult from the traceability matrix.  I'm looking forward to studying the graphs generated for each example.  :smile:

**Comment by Nathaniel-Hu** [comment]
@smiths and @JacquesCarette, I have generated another graph containing all of the data component relationships (see below links). However, it is very difficult to read/comprehend with clarity in its current state. I currently do not see any meaningful ways in which it can be split up into smaller "subgraphs". Do any of you have any suggestions for how the graph could be subdivided?

Traceability Graph 3 (with long names used):
https://github.com/JacquesCarette/Drasil/blob/Nathaniel-GenTraceGraph-%23801/People/Nathaniel/TraceGrGen3.png

Traceability Graph 3A (with short names used):
https://github.com/JacquesCarette/Drasil/blob/Nathaniel-GenTraceGraph-%23801/People/Nathaniel/TraceGrGen3A.png

**Comment by JacquesCarette** [comment]
How did you do the layout of that graph?  Did you use `dot` ? What options did you give `dot` ?

**Comment by Nathaniel-Hu** [comment]
For the graph layout, I used `rankdir=BT` (rank direction is from bottom to top) and `splines=ortho` (for orthographic relationship lines between data components). I tried playing around with `newrank`, to try and get all of the assumption data components to the bottom.  However, I haven't been able to get it to work yet.

**Comment by JacquesCarette** [comment]
What happens if you use 'raw' dot, with no directives?

**Comment by Nathaniel-Hu** [comment]
I'm not sure what you mean by use 'raw' dot, with no directives.

**Comment by JacquesCarette** [comment]
```
dot -Tpng TraceMatAvsAll_Graph.dot >TraceMatAvsAll_Graph.png
```
Switching to your branch...  If I edit `TraceGrGen3.dot` and remove all 3 lines of the `graph` declarations, I think the results are better.

Also, I feel like the arrows might be backwards. We want a "depends on" graph. You need to reverse the dependency of each of the arrows.

**Comment by JacquesCarette** [comment]
Actually, with more experimentation, if I keep things as is, remove `splines=ortho`, add "`edges [dir="back"] ;` the result seems best.

**Comment by Nathaniel-Hu** [comment]
Okay then, thanks! I'll keep that in mind when writing the code used to generate the graph .dot file. Things may change depending upon how the relationship data from the traceability matrices is obtained (whether data components "contain" all of their "dependents", or vice versa).

**Comment by smiths** [comment]
With respect to the Traceability graph example pointed to in comment:

https://github.com/JacquesCarette/Drasil/issues/801#issuecomment-628823763

I like this figure better than the previous one.  Is it generated by dot?  When I run dot, the result doesn't look like this.

If we can draw the traceability graph like a hierarchy, with levels, that would be very helpful.  It helps me to see the relationships.  As it is right now, some arrows are going up, and others down.  It would be easier to understand if it was arranged so that they all went in the same direction.

It might be too difficult to do, but it would be great to see a version of the dependency graph that follows this figure:

https://gitlab.cas.mcmaster.ca/SEforSC/se4sc/-/blob/git-svn/Writing/SIAM_CS19_Poster/RelationsBetweenTM_GD_IM_DD_A.pdf

The TMs, GDs and IMs are on top, the next "level" is data definitions and the "bottom" is Assumptions.  The relationship given explicitly in the SRS is the "reference" or "depends on" relation.  The refinement relationship is implicit.  A view of the traceability graph with Instance Models on top, then General Definitions, then Theoretical Models, then Data Definitions, then Assumptions would be nice.

I also see some disconnected components.  The disconnected functional requirements seem fine, but I'm surprised to see the DD: aspectRatio is disconnected.  If that is real, I think it points to a mistake in the documentation.  If we don't use a DD, then it shouldn't really be included in the documentation.  (These pictures really do have an important purpose for understanding the SRS document.)

There are a lot of parts that show up in the full traceability graph.  It would be great if displaying the graph was configurable.  A graph that just shows the TMs, GDs and IMs wouldn't be too big.  A graph that just shows TMs, DDs and As wouldn't be too big.  Ideally it would be fully configurable, so that the user can pick exactly what categories of data they want to show the relationship between.

**Comment by Nathaniel-Hu** [comment]
@smiths and @JacquesCarette, I have generated two more traceability graphs, each containing the rough hierarchy mentioned by Dr. Smith. The only differences are the spacing between levels/nodes, and the labelled/unlabelled subgraph clusters (see below links). Are there any changes that either of you would suggest that I make to either graph?

Sample 2 Traceability Graph (with labelled subgraph clusters/default node/level spacing):
https://github.com/JacquesCarette/Drasil/blob/Nathaniel-GenTraceGraph-%23801/People/Nathaniel/sampleTG2.png

Sample 3 Traceability Graph (with unlabelled subgraph clusters/adjusted node/level spacing):
https://github.com/JacquesCarette/Drasil/blob/Nathaniel-GenTraceGraph-%23801/People/Nathaniel/sampleTG3.png

**Comment by JacquesCarette** [comment]
I definitely prefer TG3 - it's almost readable...

**Comment by smiths** [comment]
@Nathaniel-Hu, as we discussed on Monday, I also prefer TG3.  As we further discussed, it would be nice to have a general function that can return different traceability graphs.  We could provide a list of chunk type and then generate the graph that includes these types.  This would let us generate smaller graphs that would be easier to read.

**Comment by Nathaniel-Hu** [comment]
@JacquesCarette and @smiths,
Just to clarify, where should I direct my scripts to output the `dot` files to?

**Comment by smiths** [comment]
@Nathaniel-Hu, if I understand your question, I think that the `dot` files should go in the `SRS` folder for each of the examples when they are generated.

**Comment by Nathaniel-Hu** [comment]
Ok then. Thanks!

**Comment by Nathaniel-Hu** [comment]
The progress work done so far that contributes to this issue can be found in branches Nathaniel-Issue#801 and Issue#801.

**Comment by Ant13731** [comment]
Since I'm a little bit more familiar with how graphviz works, I think I might be able to continue Nathaniel's work. Is there anything new about this issue that I should know before starting?

**Comment by JacquesCarette** [comment]
I've reread the thread - no, I think it's good to continue from where this left off.

**Comment by Ant13731** [comment]
I'm not sure if I can get the sorting nodes part down (at least probably not right away), but I'm hoping to make an autogen version of the traceability matrices currently in the examples. I've put down a few ideas in the latest commit for the Issue#801 branch

**Comment by Ant13731** [comment]
I was looking through some old branches and I think [this branch](https://github.com/JacquesCarette/Drasil/tree/Nathaniel-GenTraceGraph-%23801) could probably be deleted or merged safely. I think all of Nathaniel's work was done in the `People` so it shouldn't conflict, but on the other hand it's probably a little outdated now.
